src/savers/gif/tvgGifSaver.cpp

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2023 the ThorVG project. All rights reserved.
+
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "gif.h"
+#include "tvgGifSaver.h"
+
+#ifdef _WIN32
+    #include <malloc.h>
+#elif defined(__linux__)
+    #include <alloca.h>
+#else
+    #include <stdlib.h>
+#endif
+
+/************************************************************************/
+/* Internal Class Implementation                                        */
+/************************************************************************/
+
+bool GifSaver::makeFrame(Animation* animation, const string& path, uint32_t width, uint32_t height, uint32_t fps)
+{
+    std::unique_ptr<tvg::SwCanvas> canvas = nullptr;
+    canvas = tvg::SwCanvas::gen();
+    if (canvas == nullptr) { 
+        return false;
+    }
+
+    buffer = (uint32_t*)realloc(buffer, sizeof(uint32_t) * width * height);
+    canvas->push(tvg::cast<tvg::Picture>(animation->picture()));
+    auto sw_canvas = reinterpret_cast<tvg::SwCanvas*>(canvas.get());
+    sw_canvas->target(buffer, width, width, height, tvg::SwCanvas::ABGR8888); 
+
+    GifWriter handle;
+    GifBegin(&handle, path.c_str(), width, height, 2);
+
+    float totalFrame = animation->totalFrame();
+    uint32_t frameDelay = 0;
+    if (fps != 0) {
+        frameDelay = static_cast<uint32_t>((animation->duration() / fps) * 100.0f);
```

Comment:
Does this work correctly?

It suppose to have something like this code...

```
void GifSaver::run(unsigned tid)
{
    auto canvas = tvg::SwCanvas::gen();
    if (!canvas) return;

    //Do not share the memory pool since this canvas could be running on a thread.
    canvas->mempool(SwCanvas::Individual);
    
    auto w = static_cast<uint32_t>(vsize[0]);
    auto h = static_cast<uint32_t>(vsize[1]);

    buffer = (uint32_t*)realloc(buffer, sizeof(uint32_t) * w * h);
    canvas->target(buffer, w, w, h, tvg::SwCanvas::ABGR8888);
    canvas->push(tvg::cast<tvg::Picture>(animation->picture()));

    //use the default fps
    if (mathZero(fps)) {
        fps = (animation->totalFrame() / animation->duration());
    }

    auto delay = (fps / 60.0f);

    GifWriter writer;
    GifBegin(&writer, path, w, h, uint32_t(delay));

    auto duration = animation->duration() * 1000.0f;

    for (auto p = 0.0f; p < duration; p += delay) {
        auto frameNo = animation->totalFrame() * (p / duration);
        animation->frame(frameNo);
        canvas->update();
        if (canvas->draw() == tvg::Result::Success) { 
            canvas->sync();
        }
        GifWriteFrame(&writer, reinterpret_cast<uint8_t*>(buffer), w, h, uint32_t(delay));
    }

    GifEnd(&writer);
}
```

Reply:
> ```
>     auto duration = animation->duration() * 1000.0f;
> ```

Since GifWriteFrame is 1/100 second, I modified it to *100.0.f.